### PR TITLE
feat: See similar proposals

### PR DESCRIPTION
**Pull request by VAIBHAV-25**
https://github.com/polkassembly/polkassembly/assets/55956316/6155e3fa-200a-4cce-b4f8-0cea6490c1ef

Please Note:-

field like title, description, tags and topic will not available rightnow because their value comes from production's firebase.

### CODE REVIEW

File: pages/api/v1/listing/get-similar-posts.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+		variables: postsVariables
+	});
+	let posts: any;
+	const subsquidData = subsquidRes.data.proposals;
+	const activePostIds = subsquidData.map((proposal: any) => proposal.index);
+	let onChainCollRef;
+	if (tags && activePostIds && topicId) {
+		onChainCollRef = postsByTypeRef(network, strProposalType as ProposalType);
+		let postsSnapshotArr;
+		if (tags.lenght > 0) {
```
Comment (rajdeep7Singh) [inline pages/api/v1/listing/get-similar-posts.ts:59]: should be `length`

File: src/components/Listing/Tracks/TrackListingCard.tsx [hunk 2]
```diff
@@ -31,6 +31,7 @@ export enum CustomStatus {
 }
 
 const TrackListingCard = ({ className, posts, trackName }: Props) => {
+	console.warn(posts);
```
Comment (rajdeep7Singh) [inline src/components/Listing/Tracks/TrackListingCard.tsx:34]: remove console

File: src/queries.ts [hunk 3]
```diff
@@ -1706,3 +1706,156 @@ query VotesHistoryByVoter($type_eq: VoteType = ReferendumV2, $voter_in: [String!
...
+    }
+    type
+    createdAt
+    updatedAt
+    trackNumber
+  }
+}`;
+
+export const GET_PROPOSAL_ALLIANCE_ANNOUNCEMENT = `query getAllianceAnnouncements( $limit: Int = 50, $offset: Int = 0, $type_in:[AnnouncementType!] ) {
+  announcements(limit: $limit, offset: $offset, where:{status_in:[Rejected,Executed,TimedOut, Approved, Cancelled,ConfirmStarted, ConfirmAborted], type_in: $type_in}) {
```
Comment (rajdeep7Singh) [inline src/queries.ts:1728]: should be `status_not_in`

File: pages/api/v1/listing/get-similar-posts.ts [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	const postsVariables: any = {
+		type_eq: getSubsquidProposalType(proposalType as any)
+	};
+	const subsquidRes = await fetchSubsquid({
+		network,
+		query,
+		variables: postsVariables
+	});
+	let posts: any;
+	const subsquidData = subsquidRes.data.proposals;
```
Comment (rajdeep7Singh) [inline pages/api/v1/listing/get-similar-posts.ts:53]: null checks. use `?`

File: pages/api/v1/listing/get-similar-posts.ts [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	const subsquidData = subsquidRes?.data?.proposals;
+	const activePostIds = subsquidData.map((proposal: any) => proposal.index);
+	let onChainCollRef;
+	if (tags && activePostIds && topicId) {
+		onChainCollRef = postsByTypeRef(network, strProposalType as ProposalType);
+		let postsSnapshotArr;
+		if (tags.length > 0) {
+			postsSnapshotArr = await onChainCollRef.where('tags', 'array-contains-any', tags).get();
+		} else {
+			postsSnapshotArr = await onChainCollRef.get();
```
Comment (rajdeep7Singh) [inline pages/api/v1/listing/get-similar-posts.ts:62]: No need to iterate over all posts.

**Comment by rajdeep7Singh** [comment]
Please check on Algorithm with product. Multiple loops can be avoided

File: pages/api/v1/listing/get-similar-posts.ts [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+						};
+					}
+				}
+			});
+			posts = await Promise.all(postsPromise);
+		}
+	}
+
+	let result: any = [];
+	if (subsquidRes['data'].proposal && posts) {
```
Comment (rajdeep7Singh) [inline pages/api/v1/listing/get-similar-posts.ts:104]: change this

File: pages/api/v1/listing/get-similar-posts.ts [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+							  }
+							: getTopicFromType(strProposalType as ProposalType),
+						user_id: docData?.user_id || 1,
+						username: docData?.username
+					};
+				}
+			}
+		});
+		postTopics = await Promise.all(postsTopicPromise);
+		if (subsquidRes['data'].proposal && postTopics) {
```
Comment (rajdeep7Singh) [inline pages/api/v1/listing/get-similar-posts.ts:166]: same

File: pages/api/v1/listing/get-similar-posts.ts [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+		}
+	}
+
+	if (result.length < 3) {
+		result = subsquidData.map((proposal: any) => {
+			// console.log(`ids from non matches -> ${proposal.index}`);
+			return proposal.index;
+		});
+	}
+	result = result.filter((number: any) => number !== postId);
```
Comment (rajdeep7Singh) [inline pages/api/v1/listing/get-similar-posts.ts:183]: Avoid this loop. Don't fetch the current referenda from firestore and subsquid

Comment (VAIBHAV-25) [reply pages/api/v1/listing/get-similar-posts.ts:183]: sir without this, joh proposal open hai uski id bhi similar proposal ki ids wli array m store ho jyegi

File: pages/api/v1/listing/get-similar-posts.ts [hunk 9]
```diff
@@ -211,6 +211,9 @@ const handler: NextApiHandler<any | MessageType> = async (req, res) => {
 				} else {
 					timeline = getTimeline(post?.group?.proposals, isStatus) || [];
 				}
+				if (post.status === 'DecisionDepositPlaced') {
+					post.status = 'Deciding';
```
Comment (rajdeep7Singh) [inline pages/api/v1/listing/get-similar-posts.ts:215]: This is wrong

File: pages/api/v1/listing/get-similar-posts.ts (hunk 8 above)
Comment (rajdeep7Singh) [reply pages/api/v1/listing/get-similar-posts.ts]: use not_eq: postId in queries from subsquid/firestore to exclude current proposal

**Pull request merged**